docs/react/migration-v4.zh-CN.md

```diff
@@ -3,9 +3,126 @@ order: 7
 title: 从 v3 到 v4
 ---
 
-我们提供了一个 codemod cli 工具 [@ant-design/codemod-v4](https://github.com/ant-design/codemod-v4) 以帮助你快速升级到 v4 版本。
+本文档将帮助你从 antd `3.x` 版本升级到 antd `4.x` 版本，如果你是 `2.x` 或者更老的版本，请先参考之前的[升级文档](https://github.com/ant-design/ant-design/blob/2adf8ced24da7b3cb46a3475854a83d76a98c536/CHANGELOG.zh-CN.md#300)升级到 3.x。
 
-## 使用
+## 升级准备
+
+1. 请先升级到 3.x 的最新版本，按照控制台 warning 信息移除/修改相关的 API。
+2. 升级项目 React 16.12.0 以上。
+   - 如果你仍在使用 React 15，请参考[React 16 升级文档](https://reactjs.org/blog/2017/09/26/react-v16.0.html#breaking-changes)
+   - 其余 React 16 废弃生命周期 API 请参考 [迁移导引](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path)
+
+## 4.0 有哪些不兼容的变化
+
+### 设计规范调整
+
+- 全局阴影改变，调整为三层阴影区分控件层次关系。
+- 基础圆角调整，由`4px` 改为 `2px`。
+- 气泡确认框中图标的使用改变，由问号改为感叹号。
+- 部分组件选中颜色统一改为 `@blue-1: #E6F7FF`，对应 `hover` 颜色改为 `@gray-2: #FAFAFA`。
+- 报错色色值调整，由 `#F5222D` 改为 `#FF4D4F`。
+- 中性色使用规则变化：
+  1. 分割线颜色明度降低，由 `hsv(0, 0, 91%)` 改为 `hsv(0, 0, 94%)`。
+  2. 在深色背景下使用规则变化。
+- 灰色色板更新：
+  1. `@gary-4` 色值改变，由 `#E8E8E8` 改为 `#F0F0F0`。
+  2. 新增三个灰色。
+
+### 兼容性调整
+
+- IE 最低支持版本为 IE 11
+- React 最低支持版本为 React 16.9
+
+#### 移除废弃的 API
+
+- 移除了 LocaleProvider，请使用 `ConfigProvider` 替代。
+- 移除了 Mention，请使用 `Mentions` 替代。
+- 移除了 Alert 的 `iconType` 属性，请使用 `icon` 替代。
+- 移除了 Modal.xxx 的 `iconType` 属性，请使用 `icon` 替代。
+- 移除了 Form.create 方法，`form` 现可由 `Form.useForm` 获取。
+- 移除了 Form.Item 的 `id` 属性，请使用 `htmlFor` 替代。
+- 移除了 Typography 的 `setContentRef` 属性，请使用 `ref` 替代。
+- 移除了 TimePicker 的 `allowEmpty` 属性，请使用 `allowClear` 替代。
+- 移除了 Tag 的 `afterClose` 属性，请使用 `onClose` 替代。
+- 移除了 Card 的 `noHovering` 属性，请使用 `hoverable` 替代。
+- 移除了 Carousel 的 `vertical` 属性，请使用 `dotPosition` 替代。
+- 移除了 Drawer 的 `wrapClassName` 属性，请使用 `className` 替代。
+- 移除了 TextArea 的 `autosize` 属性，请使用 `autoSize` 替代。
+- 移除了 Affix 的 `offset` 属性，请使用 `offsetTop` 替代。
+- 移除了 Transfer 的 `onSearchChange` 属性，请使用 `onSearch` 替代。
+- 移除了 Transfer 的 `body` 属性，请使用 `children` 替代。
+- 移除了 Select 的 `combobox` 模式，请使用 `AutoComplete` 替代。
```

Comment:
`AutoComplete`的API也变了

Reply:
指的是？

Reply:
具体我也不太清楚，升级后原来的`AutoComplete`崩了。看到有deprecation warnings，但也许只是warnings。
```
Warning: `combobox` mode not support `optionLabelProp`. Please set `value` on Option directly.
Warning: [antd: AutoComplete] `dataSource` is deprecated, please use `options` instead.
Warning: [antd: AutoComplete] You need to control style self instead of setting `size` when using customize input.
```
我看文档里新版用的`options`的数据结构和原来`dataSource`不一样，不知道这是不是原因。